src/test/java/com/amplitude/HttpCallTest.java

```diff
@@ -0,0 +1,252 @@
+        },
+        errorMsg);
+    verifyConnectionOption(connection);
+  }
+
+  @ParameterizedTest
+  @MethodSource("httpCallArguments")
+  public void testHttpCallWithRateLimitResponse(HttpCallMode httpCallMode, URL url)
+      throws IOException, AmplitudeInvalidAPIKeyException {
+    JSONObject responseObject = new JSONObject();
+    String errorMsg = "Too many requests for some devices and users";
+    responseObject.put("code", 429);
+    responseObject.put("error", errorMsg);
+    int epsThreshold = 10;
+    responseObject.put("eps_threshold", epsThreshold);
+    JSONObject throttledDevices = new JSONObject();
+    throttledDevices.put("C8F9E604-F01A-4BD9-95C6-8E5357DF265D", 11);
+    responseObject.put("throttled_devices", throttledDevices);
+    JSONObject throttledUsers = new JSONObject();
+    throttledDevices.put("datamonster@amplitude.com", 12);
+    responseObject.put("throttled_users", throttledUsers);
+    JSONArray throttledEvents = new JSONArray();
+    throttledEvents.put(0);
+    responseObject.put("throttled_events", throttledEvents);
+    HttpsURLConnection connection =
+        MockHttpsURLConnectionHelper.getMockHttpsURLConnection(429, responseObject.toString());
+    mockURLStreamHandler.setConnection(url, connection);
+
+    HttpCall httpCall = getHttpCallFromCallMode(httpCallMode);
+    List<Event> events = EventsGenerator.generateEvents(1);
+    Response response = httpCall.syncHttpCallWithEventsBuffer(events);
+    assertEquals(429, response.code);
+    assertEquals(Status.RATELIMIT, response.status);
+    assertNotNull(response.rateLimitBody);
+    assertEquals(epsThreshold, response.rateLimitBody.getInt("epsThreshold"));
+    assertEquals(
+        throttledDevices.toString(),
+        response.rateLimitBody.getJSONObject("throttledDevices").toString());
+    assertEquals(
+        throttledUsers.toString(),
+        response.rateLimitBody.getJSONObject("throttledUsers").toString());
+    assertEquals(
+        Arrays.toString(Utils.jsonArrayToIntArray(throttledEvents)),
+        Arrays.toString((int[]) response.rateLimitBody.get("throttledEvents")));
+    assertNull(response.invalidRequestBody);
+    assertEquals(errorMsg, response.error);
+    assertNull(response.successBody);
+    verifyConnectionOption(connection);
+  }
+
+  @ParameterizedTest
+  @MethodSource("httpCallArguments")
+  public void testHttpCallWithIOException(HttpCallMode httpCallMode, URL url)
+      throws IOException, AmplitudeInvalidAPIKeyException {
+    HttpsURLConnection connection = mock(HttpsURLConnection.class);
+    when(connection.getOutputStream()).thenThrow(new IOException("test IOException thrown"));
+    mockURLStreamHandler.setConnection(url, connection);
+    HttpCall httpCall = getHttpCallFromCallMode(httpCallMode);
+    List<Event> events = EventsGenerator.generateEvents(1);
+    Response response = httpCall.syncHttpCallWithEventsBuffer(events);
+    assertEquals(0, response.code);
```

Comment:
Since we change to give the TIMEOUT status when throwing IOException, should we also changed here? 
https://github.com/amplitude/Amplitude-Java/pull/26

Reply:
yeah, good catch, let me pull the changes, I guess that change will break this test

Reply:
already updated